Implement timeformat option for CSV reader

nils-dohop opened the pull request:
This implements a timeformat option for the CSV reader, which comes in useful with weirdly formatted time columns.

Includes a trivial unit test.

nils-dohop committed:
Implement timeformat option for CSV reader

This implements a timeformat option for the CSV reader, which
comes in useful with weirdly formatted time columns.

Includes a trivial unit test.

hawkfish commented on src/function/scalar/strftime_format.cpp:
```diff
@@ -1360,6 +1360,13 @@ bool StrpTimeFormat::ParseResult::TryToDate(date_t &result) {
 	return Date::TryFromDate(data[0], data[1], data[2], result);
 }
 
+bool StrpTimeFormat::ParseResult::TryToTime(dtime_t &result) {
```
Unfortunately we are missing `bool Time::TryFromTime(hh, mm, ss, us, off, result)`, so this will not fail if the time is invalid. I think we need to move this duplicated code into `time.cpp` and add a call to `IsValidTime` at the end. Among other things I'm seeing that the parsing code for timestamps does not accept `24:00:00`, which is valid. This may be more than you bargained for, so unless you want to wade in deeper, this is fine for now.

hawkfish replied:
The problem with piggybacking off the timestamp parser is that if there is an offset specified in the timestamp format, then the time can be invalid. That could produce unfortunate and possibly unpleasant results.

nils-dohop replied:
I'll take a look.

nils-dohop replied:
Fixed all issues here except `24:00:00`. That will need some more consideration.

hawkfish replied:
That turns out to not be a thing. Postgres will cast `24:00:00` from ISO-8601 but not parse it in `to_timestamp` so don't worry about it.

hawkfish commented on test/sql/copy/csv/test_date.test:
```diff
@@ -37,4 +37,10 @@ query TT
+Select * from read_csv('data/csv/custom_date.csv', header=true, timeformat='%m/%d/%Y, %-I:%-M %p', types = ['BIGINT', 'TIME'] );
+----
+123	12:33:00
+124	11:57:00
```
I wish this test had some invalid data too, but the existing test does not so it is fine for now.

nils-dohop replied:
Any tips on how to run just a subset of the unit tests? I'm afraid the docs aren't that forthcoming.

nils-dohop replied:
Added a separate file for the failing tests.

hawkfish reviewed (commented):
We need to check the parsed UTC offset is zero. Adding a test for this would be a good idea.

hawkfish reviewed (changes requested):
Please change TryToTime to check for zero offset (data[7])

hawkfish committed:
Issue #11518: TryParseTime

Clean up the temporal parsing code a bit so we can support
using it to parse strangely formatted times as in #11516.

fixes: duckdb/duckdb#11516

nils-dohop commented:
Will try to fix suggestions. May need some help though. :-)

nils-dohop committed:
Implement timeformat option for CSV reader - fixes & python

Add support for time_format in the python wrapper
read_csv function.

Fix review comments:
- Validate time in time.cpp.
- Add wrapper that validates and constructs dtime_t in time.cpp
- Validate that the offset is 0 when parsing times in strftime_format.cpp
- Add unit tests for failures, one formatting failure and one value failure in
  test_time_format_errors.test

nils-dohop committed:
Implement timeformat option for CSV reader - format fix

black on a file.

nils-dohop commented:
Python build appears to be failing due to docker issues from what I can see.

hawkfish commented:
I have a [PR](https://github.com/duckdb/duckdb/pull/11519) to clean up `strftime_format` that might make your life easier.

nils-dohop commented:
Ah, I think I implemented more or less the same thing here. Have a look.

hawkfish commented on src/common/types/time.cpp:
```diff
@@ -273,6 +273,14 @@ dtime_t Time::FromTime(int32_t hour, int32_t minute, int32_t second, int32_t mic
 }
 
+bool Time::TryFromTime(int32_t hour, int32_t minute, int32_t second, int32_t microseconds, dtime_t &result) {
+	if (!Time::IsValidTime(hour, minute, second, microseconds)) {
```
The parser guarantees this to be true. There is an asymmetry with dates because dates need to be checked for month length, leap years and so on. I'd merge my PR (it's only two files).

hawkfish replied:
What time ranges are valid depends on the context, so this is not always correct (it is not correct when using a format string for example). So let's just remove this.

hawkfish commented on src/include/duckdb/common/types/time.hpp:
```diff
@@ -37,6 +37,8 @@ class Time {
 	DUCKDB_API static string ToUTCOffset(int hour_offset, int minute_offset);
 
 	DUCKDB_API static dtime_t FromTime(int32_t hour, int32_t minute, int32_t second, int32_t microseconds = 0);
+	DUCKDB_API static bool TryFromTime(int32_t hour, int32_t minute, int32_t second, int32_t microseconds,
```
I don't think this is needed if you use my cleanup.

nils-dohop replied:
I think it's clean though in the sense that the function ends up in `time.hpp` which mirrors the structure of date and timestamp.

nils-dohop replied:
I'll merge your PR though, I guess it effectively increases performance.

nils-dohop replied:
I cherry-picked your commit, hope that works for you.

nils-dohop replied:
I left these in though, for completeness one could say, time then has the same functions as date and timestamp. I can delete them though if you think that's more suitable.

hawkfish replied:
Yes please remove - dates and times are not parallel (unfortunately...)

hawkfish commented on tools/jdbc/generator.py:
```diff
@@ -17,9 +17,8 @@ def function_hook(fn: CppMethod, config):
-            'short_name': (
-                '_duckdb_jdbc_' + name.replace('Java_org_duckdb_DuckDBNative_duckdb_1jdbc_1', '').replace('1', '')
-            ),
+            'short_name': '_duckdb_jdbc_'
```
Is this change intentional?

nils-dohop replied:
It wasn't my idea, came in from black.

hawkfish reviewed (changes requested):
Getting there! This is feature work so it's probably not going to make the next bug drop on Monday, so why don't you merge in my cleanup PR (which is also tagged as feature work) so we don't collide?

hawkfish commented:
OK my PR got merged earlier, so why don't you merge main and push again?

hawkfish reviewed (changes requested):
Just remove the changes to time.* and merge from main and I think we are good to go.

nils-dohop committed:
Merge remote-tracking branch 'origin/main' into add-timeformat-option-for-csv-reader

nils-dohop committed:
Remove unneeded time functions

nils-dohop commented:
@hawkfish thanks for all the feedback. Should be good to go now.

hawkfish reviewed (approved):
Thanks - LGTM!

Mytherin commented:
@pdet can you also have a look?

pdet commented on test/sql/copy/csv/test_csv_time_format_errors.test:
```diff
@@ -0,0 +1,25 @@
+# description: Test reading csv files containing time format errors
+# group: [csv]
+
+statement error
```
Could you add 
```
statement ok
PRAGMA enable_verification
```

to the top of all sql tests and make sure the new option is properly serialized?

pdet commented on test/sql/copy/csv/test_csv_time_format_errors.test:
```diff
@@ -0,0 +1,25 @@
+Conversion Error: CSV Error on Line: 3
+Error when converting column "Date_Time".
+Could not parse string "23:67" according to format specifier "%H:%M"
+23:67
```
Could you also add tests with the `ignore_errors` option set to true?

pdet replied:
Could we also add tests with CSV files that are over one vector size (i.e., > 2048 elements).

pdet reviewed (changes requested):
Hi @nils-dohop many thanks for the PR. I think it's looking great.

I just asked for a few tests. I do expect the serializiation test to not work.

For that, you basically will need to add the new option to `src/include/duckdb/storage/serialization/nodes.json` and run `make generate-files`

I also think it would be interesting to add the time mask to the sniffer with some common mask options, but I believe that can also be added later (What do you think @Mytherin ?)

pdet commented:
Also @Mytherin should this be re-targeted to feature?

nils-dohop commented:
I'm abandoning this PR. Sorry, but things seem to have changed too much for this to have any future. I'll just work around this instead.

nils-dohop commented:
I've fixed the remaining comments, @pdet , but based on the other work that was done in the same files this PR is no longer current, if you feel like figuring out what's needed to make it work again, please feel free to merge the commits.

Tishj commented:
@nils-dohop If you're talking about the python side of things it's much easier than it was before
As explained by this PR <https://github.com/duckdb/duckdb/pull/11202>

The only thing necessary is to edit the `tools/pythonpkg/scripts/connection_methods.json` file, alongside the changes to pyconnection.hpp and pyconnection.cpp

nils-dohop commented:
@Tishj No, the C++ parts have also moved substantially.
